File: final_test.py
#!/usr/bin/env python3
"""
Final test to verify all fixes
"""

import subprocess
import os

def run_test(name, cmd):
    """Run a single test"""
    rs_dash_path = os.path.join(os.path.dirname(__file__), "target", "debug", "rs-dash")
    if os.name == 'nt':
        rs_dash_path += ".exe"
    
    print(f"\n{name}:")
    print(f"  Command: {cmd}")
    
    try:
        result = subprocess.run([rs_dash_path, "-c", cmd],
                               capture_output=True, text=True, timeout=3)
        
        if result.returncode == 0:
            print(f"  OK Exit: {result.returncode}")
        else:
            print(f"  FAIL Exit: {result.returncode}")
        
        if result.stdout.strip():
            print(f"  Output: {result.stdout.strip()}")
        
        if result.stderr.strip():
            print(f"  Error: {result.stderr.strip()}")
            
    except subprocess.TimeoutExpired:
        print(f"  ✗ TIMEOUT: Command hung (this was the bug!)")

def main():
    print("=== Final Verification of All Fixes ===")
    
    # Original issues from user
    print("\n--- Original Issues ---")
    
    # Issue 1: Variable expansion
    run_test("Issue 1: $? variable", "true; echo $?; false; echo $?")
    
    # Issue 2: Command substitution  
    run_test("Issue 2: $(command) substitution", "echo $(echo test)")
    
    # Issue 3: Pipeline (user said it hangs)
    run_test("Issue 3: Pipeline echo hello | echo world", "echo hello | echo world")
    
    # Additional tests
    print("\n--- Additional Tests ---")
    
    run_test("Special var $$", "echo $$")
    run_test("Special var $0", "echo $0")
    run_test("Variable assignment", "MYVAR=test && echo $MYVAR")
    run_test("Multiple commands", "echo a; echo b; echo c")
    run_test("Logical AND", "true && echo success")
    run_test("Logical OR", "false || echo failure")
    run_test("Complex pipeline", "echo start && echo middle | echo end")
    run_test("Nested operators", "false || echo alt | echo result")

if __name__ == "__main__":
    main()

File: final_test.rs
use std::process;

fn main() {
    println!("=== Final Priority 1 Test ===\n");
    
    let tests = vec![
        // Parameter expansion tests
        ("Basic parameter", "VAR=hello; echo ${VAR}", "hello"),
        ("Default value", "echo ${UNDEF:-default}", "default"),
        ("Assign default", "echo ${UNDEF:=assigned}; echo $UNDEF", "assigned\nassigned"),
        ("String length", "VAR=hello; echo ${#VAR}", "5"),
        ("Pattern suffix", "VAR=filename.txt; echo ${VAR%.txt}", "filename"),
        ("Pattern prefix", "VAR=filename.txt; echo ${VAR#file}", "name.txt"),
        ("Pattern substitution", "VAR=hello world; echo ${VAR/world/everyone}", "hello everyone"),
        ("Case upper", "VAR=hello; echo ${VAR^^}", "HELLO"),
        ("Case lower", "VAR=HELLO; echo ${VAR,,}", "hello"),
        
        // Arithmetic expansion tests
        ("Arithmetic basic", "echo $((1 + 2))", "3"),
        ("Arithmetic precedence", "echo $((1 + 2 * 3))", "7"),
        ("Arithmetic bitwise", "echo $((5 & 3))", "1"),
        ("Arithmetic comparison", "echo $((2 > 1))", "1"),
        ("Arithmetic ternary", "echo $((1 ? 100 : 200))", "100"),
        
        // Positional parameters tests
        ("Positional count", "echo $#", "3"),
        ("Positional first", "echo $1", "arg1"),
        ("Positional second", "echo $2", "arg2"),
        ("Positional all", "echo $@", "arg1 arg2 arg3"),
        ("Shell name", "echo $0", "rs-dash"),
        
        // Special parameters tests
        ("Exit status success", "true; echo $?", "0"),
        ("Exit status failure", "false; echo $?", "1"),
        ("PID", "echo $$", ""), // Just check it doesn't crash
    ];
    
    let mut passed = 0;
    let mut failed = 0;
    
    for (name, cmd, expected) in tests {
        let output = process::Command::new("target/debug/rs-dash")
            .arg("-c")
            .arg(cmd)
            .arg("arg1")
            .arg("arg2")
            .arg("arg3")
            .output();
        
        match output {
            Ok(output) => {
                let stdout = String::from_utf8_lossy(&output.stdout).trim().to_string();
                let stderr = String::from_utf8_lossy(&output.stderr).trim().to_string();
                
                // For PID test, we just check it doesn't crash
                let success = if name == "PID" {
                    output.status.success() && stderr.is_empty()
                } else {
                    stdout == expected && output.status.success()
                };
                
                if success {
                    println!("✓ {}: PASS", name);
                    passed += 1;
                } else {
                    println!("✗ {}: FAIL", name);
                    println!("  Command: {}", cmd);
                    println!("  Expected: {}", expected);
                    println!("  Got: {}", stdout);
                    if !stderr.is_empty() {
                        println!("  Stderr: {}", stderr);
                    }
                    failed += 1;
                }
            }
            Err(e) => {
                println!("✗ {}: ERROR - {}", name, e);
                failed += 1;
            }
        }
    }
    
    println!("\n=== Summary ===");
    println!("Total tests: {}", passed + failed);
    println!("Passed: {}", passed);
    println!("Failed: {}", failed);
    
    if failed == 0 {
        println!("\nAll Priority 1 tests passed! ✓");
    } else {
        println!("\nSome tests failed.");
        std::process::exit(1);
    }
}